modules/dag-l0/src/main/scala/org/tessellation/dag/l0/domain/snapshot/programs/GlobalSnapshotEventCutter.scala

```diff
@@ -0,0 +1,134 @@
+        startingEvents.tailRecM { events =>
+          for {
+            eventToRemove <- MonadThrow[F].fromOption(iterator.nextOption(), new IllegalStateException("No GlobalSnapshotEvent events"))
+            reducedEvents = removeOneEvent(events, eventToRemove)
+            cutAgain <- needsCutting(reducedEvents)
+          } yield Either.cond(!cutAgain, reducedEvents, reducedEvents)
+        }
+
+      def removeOneEvent(events: GlobalPeerIdEventMap, eventToRemove: PeerIdEvent): GlobalPeerIdEventMap =
+        events.updatedWith(eventToRemove.id) {
+          case None => None
+          case Some(globalPeerIdEvent) =>
+            val globalSnapshotEvent = eventToRemove.toGlobalSnapshotEvent
+            val updatedEvents = globalPeerIdEvent.filterNot { case (_, gse) => gse == globalSnapshotEvent }
+            Option.when(updatedEvents.nonEmpty)(updatedEvents)
+        }
+
+      trait PeerIdEvent {
+        def id: PeerId
+        def fee: Long
+        def hash: Hash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent
+      }
+
+      val ordering: Ordering[PeerIdEvent] = Ordering.by[PeerIdEvent, (Long, Hash)](e => (e.fee, e.hash))
+
+      case class PeerIdStateChannelEvent(id: PeerId, event: StateChannelEvent) extends PeerIdEvent {
+        val fee: Long = event.snapshotBinary.value.fee.value.value
+        def hash: Hash = event.snapshotBinary.lastSnapshotHash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent = event.asLeft[DAGEvent]
+      }
+
+      case class PeerIdDAGBlockEvent(id: PeerId, event: Hashed[Block]) extends PeerIdEvent {
+        val fee: Long = event.signed.value.transactions.toIterable.map(_.value.fee.value.value).sum
+        def hash: Hash = event.hash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent = event.signed.asRight[StateChannelEvent]
+
+        def isMyChild(that: PeerIdDAGBlockEvent): Boolean =
+          that.event.signed.value.parent.contains_(event.ownReference)
+      }
+
+      class OrderedEvents(sortedStateEvents: List[PeerIdStateChannelEvent], hashedEvents: List[PeerIdDAGBlockEvent]) {
+        def next: Option[(PeerIdEvent, OrderedEvents)] = {
+          val (parents, nonParents) =
+            hashedEvents.foldLeft((List.empty[PeerIdDAGBlockEvent], List.empty[PeerIdDAGBlockEvent])) {
+              case ((parents, nonParents), p) if hashedEvents.exists(p.isMyChild) => (p :: parents, nonParents)
+              case ((parents, nonParents), p)                                     => (parents, p :: nonParents)
+            }
+
+          (sortedStateEvents, nonParents.sorted(ordering)) match {
+            case (Nil, Nil) =>
+              none
+            case (h :: t, Nil) =>
+              (h, new OrderedEvents(t, Nil)).some
+            case (Nil, h :: t) =>
+              (h, new OrderedEvents(Nil, t ++ parents)).some
+            case (h1 :: t1, h2 :: t2) =>
+              if (h1.fee < h2.fee || (h1.fee == h2.fee && sortedStateEvents.length >= hashedEvents.length))
```

Comment:
> StateChannelEvent removed before DAGEvent events in case of a tie?

This is fine, it's a rare situation for tie and only goal is to have rough equality.